go-controller/pkg/metrics/recorder/recorder.go

```diff
@@ -0,0 +1,489 @@
+	Buckets:   prometheus.ExponentialBuckets(.01, 2, 18),
+})
+
+const (
+	MetricOvnkubeNamespace       = "ovnkube"
+	MetricOvnkubeSubsystemMaster = "master"
+	portBindingTable             = "Port_Binding"
+	nbGlobalTable                = "NB_Global"
+	queueCheckPeriod             = time.Millisecond * 500
+	// queue operations
+	// pod setup latency monitor
+	addPortBinding operation = iota
+	updatePortBinding
+	addPod
+	cleanPod
+	addLogicalSwitchPort
+	// config duration monitor - queue operations
+	configStart
+	configEnd
+	// pod setup latency monitor timestamp metadata
+	// pod event first handled by OVN-Kubernetes control plane
+	firstSeen timestampType = iota
+	// OVN-Kubernetes control plane created Logical Switch Port in northbound database
+	logicalSwitchPort
+	// port binding seen in OVN-Kubernetes control plane southbound database libovsdb cache
+	portBinding
+	// port binding with updated chassis seen in OVN-Kubernetes control plane southbound database libovsdb cache
+	portBindingChassis
+	// disable config duration monitor if cluster node limit is reached to protect SB DB.
+	configDurationNodeLimit         = 50
+	configDurationNodeRefreshPeriod = 1 * time.Hour
+	configDurationStopMsg           = "Stopping configuration duration monitor due to node count greater than %d " +
+		"to protect southbound database from overload"
+)
+
+type (
+	timestampType int
+	operation     int
+)
+
+type podRecord struct {
+	timestamp time.Time
+	timestampType
+}
+
+// pItem is the data structure for pod setup latency monitor queue
+type pItem struct {
+	op        operation
+	timestamp time.Time
+	old       model.Model
+	new       model.Model
+	podUID    kapimtypes.UID
+}
+
+type Pod struct {
+	records map[kapimtypes.UID]*podRecord
+	queue   workqueue.Interface
+}
+
+type nbCfgRecord struct {
+	// timestamp of when ovn-kube master process send the Tx to OVN NB.
+	timestamp int
+	nbcfg     int
+}
+
+// cItem is the data structure for config duration monitor queue
+type cItem struct {
+	op        operation
+	timestamp int
+	cfg       int
+}
+
+type Config struct {
+	enabled bool
+	nbCfg   int64
+	records []*nbCfgRecord
+	queue   workqueue.Interface
+}
+
+type Recorder struct {
+	Pod    Pod
+	Config Config
+}
+
+var r *Recorder
+
+func NewRecorder() *Recorder {
+	r = &Recorder{}
+	return r
+}
+
+// GetRecorder returns recorder. Recorder is active if it is not nil and must be checked.
+func GetRecorder() (*Recorder, error) {
+	if r == nil {
+		return nil, fmt.Errorf("recorder has not been created")
+	}
+	return r, nil
+}
+
+//Run monitors pod setup latency and the duration for OVN-Kube master to apply configuration to all nodes.
+func (cpr *Recorder) Run(sbClient, nbClient libovsdbclient.Client, kube kube.Interface, stop <-chan struct{}) {
+	cpr.Pod.run(sbClient, stop)
+	if config.Monitoring.EnableConfigDurationMonitor {
+		// OVN-Kube master process sends txs to OVN NB DB and if config duration monitor is enabled, it also adds
```

Comment:
Killer comment. Awesome. Thanks for writing this.

Reply:
All credit to Dumitru Ceara for explaining it to me.